packages/terminal/src/browser/terminal-quick-open-service.ts

```diff
@@ -0,0 +1,174 @@
+            label: 'Open New Terminal',
+            iconClass: 'fa fa-plus',
+            run: this.doCreateNewTerminal(),
+            groupLabel: undefined,
+            showBorder: !!terminalItems.length
+        });
+        terminalItems.push(createNewTerminalItem);
+
+        acceptor(terminalItems);
+        return;
+    }
+
+    /**
+     * Compare two terminal widgets by label. If labels are identical, compare by the widget id.
+     * @param a `TerminalWidget` for comparison
+     * @param b `TerminalWidget` for comparison
+     */
+    protected compareItems(a: TerminalWidget, b: TerminalWidget): number {
+        const normalize = (str: string) => str.trim().toLowerCase();
+
+        if (normalize(a.title.label) !== normalize(b.title.label)) {
+            return normalize(a.title.label).localeCompare(normalize(b.title.label));
+        } else {
+            return normalize(a.id).localeCompare(normalize(b.id));
+        }
+    }
+
+    /**
+     * Get the function that can create a new terminal.
+     * @param {TerminalWidget} widget - the terminal widget to be opened.
+     * @returns Function that would create a new terminal if mode === QuickOpenMode.OPEN.
+     */
+    protected doCreateNewTerminal(): (mode: QuickOpenMode) => boolean {
+        return (mode: QuickOpenMode) => {
+            if (mode !== QuickOpenMode.OPEN) {
+                return false;
+            }
+            this.commandRegistry.executeCommand(TerminalCommands.NEW.id);
+            return true;
+        };
+    }
+
+    /**
+     * Convert the terminal widget to the quick open item.
+     * @param {TerminalWidget} widget - the terminal widget.
+     * @returns The quick open group item.
+     */
+    protected async toItem(widget: TerminalWidget): Promise<QuickOpenGroupItem<QuickOpenItemOptions>> {
+        const options: QuickOpenGroupItemOptions = {
+            label: widget.title.label,
+            description: widget.id,
+            tooltip: widget.title.label,
+            hidden: false,
+            run: this.getRunFunction(widget),
+            groupLabel: undefined,
+            showBorder: false
+        };
+        return new QuickOpenGroupItem<QuickOpenGroupItemOptions>(options);
+    }
+
+    /**
+     * Get the function that can open the editor file.
+     * @param {TerminalWidget} widget - the terminal widget to be opened.
+     * @returns Function that would open the terminal if mode === QuickOpenMode.OPEN.
+     */
+    protected getRunFunction(widget: TerminalWidget): (mode: QuickOpenMode) => boolean {
+        return (mode: QuickOpenMode) => {
+            if (mode !== QuickOpenMode.OPEN) {
+                return false;
+            }
+            this.terminalService.open(widget);
+            return true;
+        };
+    }
+}
+
+/**
+ * TODO: merge it to TerminalFrontendContribution
+ */
+@injectable()
+export class TerminalQuickOpenContribution implements CommandContribution, QuickOpenContribution {
```

Comment:
Why is not it a part of terminal frontend contribution? I cannot get a logic to introduce a new contribution for each command. One extension should have one entry contribution. We need to have something like the design guideline to collect such rules. or add it to the coding guidline.

Reply:
Doing so would cause the circular dependency issue. The `TerminalQuickOpenService` injects `TerminalService`, which is implemented by `TerminalFrontendContribution`. If `TerminalFrontendContribution` class incluced the code for registering command, it would need to inject the `TerminalQuickOpenService`.

Reply:
> Why is not it a part of terminal frontend contribution? I cannot get a logic to introduce a new contribution for each command. One extension should have one entry contribution. We need to have something like the design guideline to collect such rules. or add it to the coding guidline.

Ideally it should, and putting the coed into a separate class is temporary due to the existing issue.  We are hoping a refactoring (maybe the `TerminalFrontendContribution` should not implement the terminal service) can be done to avoid this type of issues in the Terminal package.

Reply:
ok, i see